From: Ryan Roberts <ryan.roberts@arm.com>
Subject: mm-madvise-avoid-split-during-madv_pageout-and-madv_cold-fix
Date: Wed, 3 Apr 2024 18:17:56 +0100

arch_enter_lazy_mmu_mode() needs to be before the continue

Signed-off-by: Ryan Roberts <ryan.roberts@arm.com>
Cc: Barry Song <21cnbao@gmail.com>
Cc: Barry Song <v-songbaohua@oppo.com>
Cc: Chris Li <chrisl@kernel.org>
Cc: David Hildenbrand <david@redhat.com>
Cc: Gao Xiang <xiang@kernel.org>
Cc: "Huang, Ying" <ying.huang@intel.com>
Cc: Kefeng Wang <wangkefeng.wang@huawei.com>
Cc: Lance Yang <ioworker0@gmail.com>
Cc: Matthew Wilcox (Oracle) <willy@infradead.org>
Cc: Michal Hocko <mhocko@suse.com>
Cc: Yang Shi <shy828301@gmail.com>
Cc: Yu Zhao <yuzhao@google.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/madvise.c |    5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

--- a/mm/madvise.c~mm-madvise-avoid-split-during-madv_pageout-and-madv_cold-fix
+++ a/mm/madvise.c
@@ -486,10 +486,9 @@ restart:
 					pte_offset_map_lock(mm, pmd, addr, &ptl);
 				if (!start_pte)
 					break;
-				if (err)
-					continue;
 				arch_enter_lazy_mmu_mode();
-				nr = 0;
+				if (!err)
+					nr = 0;
 				continue;
 			}
 		}
_
